### PR TITLE
Add new options 'match' and 'values'.  Add new op 'get'.

### PR DESCRIPTION
**Pull request by ncabatoff**
New set mode: set path=config match={} values={}

  Find uci sections based on path and match, modify their
  properties according to values.

```
  # Modify settings for pre-existing wireless interface with device radio0.
  - name: configure wireless
    uci:
      path: wireless
      match:
        device: radio0
        type: wifi-iface
      values:
        ssid: myssid
        encryption: psk2
        key: "secret passphrase"
```

New op: get path=config[.section] [type=t] [match=m]

  Used with register var x, outputs to x.result['values']
  (x.result.values won't work.)  This can be used for conditions
  in subsequent tasks.

```
  # See if there already exists a forwarding config with dest=myvpn and src=lan
  - name: Get myvpn zone forwarding
    uci: op=get path=firewall type=forwarding match="dest=myvpn src=lan"
    register: myvpnfw
  # Create the section if it doesn't exist
  - name: Add myvpn zone forwarding
    uci: autocommit=false path=firewall type=forwarding
    when: not myvpnfw.result['values']
  # Populate the section if it didn't exist
  - name: Setup myvpn zone forwarding
    uci: autocommit=false path=firewall.@forwarding[-1].{{ item.key }} value={{ item.value}}
    with_dict:
      dest: myvpn
      src: lan
    when: not myvpnfw.result['values']
```

### CODE REVIEW

File: src/uci.lua [hunk 1]
```diff
@@ -175,7 +196,16 @@ function set_value(module)
...
 	local res
-	if not sec then
+	if nil ~= p["match"] then
+		local message = {
+			config=conf,
+			values=p["values"],
+			match=p["match"]
+		}
+		-- TODO: how can we properly report whether a changed happend in this case?
+		-- Get the entire config before and after to compare?
```
Comment (noctux) [inline src/uci.lua]: UCI has the option to query changes (globally, or for specific sections) and exposes it via:
```
ubus call uci changes "{'config': 'network'}"
```
(you can test this via the ubus binary on your openwrt box to get a feeling of the return values)
I think this is the most promising way here, to do a diff of the changes before and after the set.
Naturally, it has to be called before the commit operation, though as we usually call `docommit` just before `module_exit`, this should not be an issue.
Cons: We have to write a differ for lua tables?

**Comment by noctux** [comment]
Hey,

thank you for the pull request!
I generally like the syntax. Personally, I've always tackled the problem by gathering the facts beforehand via `ubus: cmd=facts`, and filtering on Ansible-template level and sometimes via python-filters, which is far from optimal.

Unfortunately, I'm a bit stressed until Friday, so I'll do a more complete review on the weekend.
Still I wanted to give you a thumbs up :)

On a cursory glance (and with the help of your comment), it seems like we loose the ability to correctly report the change status with this operation, right?
As I said in the review, diffing changes for the relevant path seems to be our best option. What is your take on that?

Best wishes,
Simon

**Comment by ncabatoff** [comment]
Thanks for the feedback.  I too would prefer not to have to introduce a table differ.  We know which config is being modified, so we can call uci changes specifying that config, which yields a simple output format:

```
root@wdr3500:/etc/config# ubus call uci changes '{"config": "wireless"}'
{
	"changes": [
		[
			"set",
			"radio1",
			"disabled",
			"1"
		],
		[
			"set",
			"radio1",
			"disabled",
			"0"
		]
	]
}
```

It looks like each change made is appended to the list.  I'm still testing, but I think we can do something as simple as removing the first N elements of the second call to uci changes, where N is the size of the first call's result.

**Comment by noctux** [comment]
It seems like we are indeed lucky. Multiple changes to the same value also seem to be stored as multiple updates:
```sh
ansible openwrt -i hosts -m uci -a "name=uhttpd.test.foo value=bar type="uhttpd" autocommit=false"
# ...
ansible openwrt -i hosts -m uci -a "name=uhttpd.test.foo value=baz type="uhttpd" autocommit=false"
# ...
ansible openwrt -i hosts -m ubus -a "cmd=call path=uci method=changes message={}"                                                                                  
openwrt | SUCCESS => {
    "changed": true, 
    "invocations": {
        "module_args": {
            "command": "call", 
            "message": {}, 
            "method": "changes", 
            "path": [
                "uci"
            ]
        }
    }, 
    "msg": "Called uci.changes({})", 
    "result": {
        "changes": {
            "uhttpd": [
                [
                    "set", 
                    "test", 
                    "uhttpd"
                ], 
                [
                    "set", 
                    "test", 
                    "foo", 
                    "bar"
                ], 
                [
                    "set", 
                    "test", 
                    "foo", 
                    "baz"
                ]
            ]
        }
    }
}
```

So the approach looks fine to me, well spottet. For future reference, if this assumption breaks down one day, we already have a non recursive level 1 `compare_tables` in `uci.lua`, which should be fairly easy to extend.
I did not spot any other too obvious problems in the code either and tests worked fine, so I'll merge it.

Thank you for your contribution!

**Comment by ncabatoff** [comment]
That's great, thanks very much!

**Pull request merged**